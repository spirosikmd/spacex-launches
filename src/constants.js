export const ASC = 'asc';
export const DESC = 'desc';
export const FLIGHT_NUMBER_FIELD = 'flight_number';
export const UTC_DATE_FIELD = 'launch_date_utc';
export const FIELDS = `${FLIGHT_NUMBER_FIELD},${UTC_DATE_FIELD},launch_success,upcoming,is_tentative,details,mission_name,mission_id`;
