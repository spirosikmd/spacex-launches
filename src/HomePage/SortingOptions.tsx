import React, { SyntheticEvent } from 'react';
import PropTypes from 'prop-types';
import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';
import { ASC, DESC, FLIGHT_NUMBER_FIELD, UTC_DATE_FIELD } from '../constants';

function getSortOrderOptions(sortField: string) {
  switch (sortField) {
    case FLIGHT_NUMBER_FIELD:
      return (
        <>
          <option value={ASC}>smaller to bigger</option>
          <option value={DESC}>bigger to smaller</option>
        </>
      );
    case UTC_DATE_FIELD:
      return (
        <>
          <option value={ASC}>old to new</option>
          <option value={DESC}>new to old</option>
        </>
      );
    default:
      return (
        <>
          <option value={ASC}>ascending</option>
          <option value={DESC}>descending</option>
        </>
      );
  }
}

interface SortingOptionsProps {
  sortField: string;
  sortOrder: string;
  onSortChange: (name: string, value: string) => void;
}

const SortingOptions = ({
  sortField,
  sortOrder,
  onSortChange,
}: SortingOptionsProps) => {
  const handleSortChange = (event: SyntheticEvent) => {
    const target = event.currentTarget as HTMLSelectElement;
    onSortChange(target.name, target.value);
  };

  return (
    <Grid container spacing={8}>
      <Grid item>
        <TextField
          id="sort-field"
          select
          label="Sort field"
          value={sortField}
          onChange={handleSortChange}
          SelectProps={{
            native: true,
            name: 'sortField',
          }}
          margin="normal"
          variant="outlined"
        >
          <option value={FLIGHT_NUMBER_FIELD}>flight number</option>
          <option value={UTC_DATE_FIELD}>utc date</option>
        </TextField>
      </Grid>
      <Grid item>
        <TextField
          id="sort-order"
          select
          label="Sort order"
          value={sortOrder}
          onChange={handleSortChange}
          SelectProps={{
            native: true,
            name: 'sortOrder',
          }}
          margin="normal"
          variant="outlined"
        >
          {getSortOrderOptions(sortField)}
        </TextField>
      </Grid>
    </Grid>
  );
};

SortingOptions.propTypes = {
  sortField: PropTypes.string.isRequired,
  sortOrder: PropTypes.string.isRequired,
  onSortChange: PropTypes.func.isRequired,
};

export default React.memo(SortingOptions);
