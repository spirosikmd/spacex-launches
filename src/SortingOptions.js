import React from 'react';
import PropTypes from 'prop-types';
import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';
import { ASC, DESC, FLIGHT_NUMBER_FIELD, UTC_DATE_FIELD } from './constants';

export const SortingOptions = ({ sortField, sortOrder, onSortChange }) => {
  const handleSortChange = event => {
    onSortChange(event.currentTarget.name, event.currentTarget.value);
  };

  return (
    <Grid container spacing={8}>
      <Grid item>
        <TextField
          id="sort-field"
          select
          label="Sort field"
          value={sortField}
          onChange={handleSortChange}
          SelectProps={{
            native: true,
            name: 'sortField',
          }}
          margin="normal"
          variant="outlined"
        >
          <option value={FLIGHT_NUMBER_FIELD}>flight number</option>
          <option value={UTC_DATE_FIELD}>utc date</option>
        </TextField>
      </Grid>
      <Grid item>
        <TextField
          id="sort-order"
          select
          label="Sort order"
          value={sortOrder}
          onChange={handleSortChange}
          SelectProps={{
            native: true,
            name: 'sortOrder',
          }}
          margin="normal"
          variant="outlined"
        >
          <option value={ASC}>ascending</option>
          <option value={DESC}>descending</option>
        </TextField>
      </Grid>
    </Grid>
  );
};

SortingOptions.propTypes = {
  sortField: PropTypes.string.isRequired,
  sortOrder: PropTypes.string.isRequired,
  onSortChange: PropTypes.func.isRequired,
};

export default React.memo(SortingOptions);
