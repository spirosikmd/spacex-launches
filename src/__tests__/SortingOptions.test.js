import TextField from '@material-ui/core/TextField';
import SortingOptions from '../SortingOptions';
import { DESC, FLIGHT_NUMBER_FIELD, UTC_DATE_FIELD } from '../constants';

describe('SortingOptions', () => {
  let props;

  beforeEach(() => {
    props = {
      sortOrder: DESC,
      sortField: FLIGHT_NUMBER_FIELD,
      onSortChange: jest.fn(),
    };
  });

  describe('when sorting field is flight number', () => {
    it('renders sorting options with flight number specific text', () => {
      expect(shallow(SortingOptions, props)).toMatchSnapshot();
    });
  });

  describe('when sorting field is utc date', () => {
    it('renders sorting options with date specific text', () => {
      props.sortField = UTC_DATE_FIELD;
      expect(shallow(SortingOptions, props)).toMatchSnapshot();
    });
  });

  describe('when sorting changes', () => {
    it('calls the onSortChange prop', () => {
      const wrapper = shallow(SortingOptions, props);
      const sortField = wrapper.find(TextField).at(0);
      sortField.prop('onChange')({
        currentTarget: {
          name: 'name',
          value: 'value',
        },
      });
      expect(props.onSortChange).toBeCalledWith('name', 'value');
    });
  });
});
