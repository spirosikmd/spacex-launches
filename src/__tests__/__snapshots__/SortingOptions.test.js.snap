// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortingOptions when shows upcoming renders upcoming text in the options 1`] = `
<WithStyles(Grid)
  container={true}
  spacing={8}
>
  <WithStyles(Grid)
    item={true}
  >
    <TextField
      SelectProps={
        Object {
          "name": "sortField",
          "native": true,
        }
      }
      id="sort-field"
      label="Sort field"
      margin="normal"
      onChange={[Function]}
      required={false}
      select={true}
      value="launch_date_utc"
      variant="outlined"
    >
      <option
        value="flight_number"
      >
        flight number
      </option>
      <option
        value="launch_date_utc"
      >
        utc date
      </option>
    </TextField>
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
  >
    <TextField
      SelectProps={
        Object {
          "name": "sortOrder",
          "native": true,
        }
      }
      id="sort-order"
      label="Sort order"
      margin="normal"
      onChange={[Function]}
      required={false}
      select={true}
      value="desc"
      variant="outlined"
    >
      <option
        value="asc"
      >
        old to 
        upcoming
      </option>
      <option
        value="desc"
      >
        upcoming
         to old
      </option>
    </TextField>
  </WithStyles(Grid)>
</WithStyles(Grid)>
`;

exports[`SortingOptions when sorting field is flight number renders sorting options with flight number specific text 1`] = `
<WithStyles(Grid)
  container={true}
  spacing={8}
>
  <WithStyles(Grid)
    item={true}
  >
    <TextField
      SelectProps={
        Object {
          "name": "sortField",
          "native": true,
        }
      }
      id="sort-field"
      label="Sort field"
      margin="normal"
      onChange={[Function]}
      required={false}
      select={true}
      value="flight_number"
      variant="outlined"
    >
      <option
        value="flight_number"
      >
        flight number
      </option>
      <option
        value="launch_date_utc"
      >
        utc date
      </option>
    </TextField>
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
  >
    <TextField
      SelectProps={
        Object {
          "name": "sortOrder",
          "native": true,
        }
      }
      id="sort-order"
      label="Sort order"
      margin="normal"
      onChange={[Function]}
      required={false}
      select={true}
      value="desc"
      variant="outlined"
    >
      <option
        value="asc"
      >
        smaller to bigger
      </option>
      <option
        value="desc"
      >
        bigger to smaller
      </option>
    </TextField>
  </WithStyles(Grid)>
</WithStyles(Grid)>
`;

exports[`SortingOptions when sorting field is utc date renders sorting options with date specific text 1`] = `
<WithStyles(Grid)
  container={true}
  spacing={8}
>
  <WithStyles(Grid)
    item={true}
  >
    <TextField
      SelectProps={
        Object {
          "name": "sortField",
          "native": true,
        }
      }
      id="sort-field"
      label="Sort field"
      margin="normal"
      onChange={[Function]}
      required={false}
      select={true}
      value="launch_date_utc"
      variant="outlined"
    >
      <option
        value="flight_number"
      >
        flight number
      </option>
      <option
        value="launch_date_utc"
      >
        utc date
      </option>
    </TextField>
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
  >
    <TextField
      SelectProps={
        Object {
          "name": "sortOrder",
          "native": true,
        }
      }
      id="sort-order"
      label="Sort order"
      margin="normal"
      onChange={[Function]}
      required={false}
      select={true}
      value="desc"
      variant="outlined"
    >
      <option
        value="asc"
      >
        old to 
        recent
      </option>
      <option
        value="desc"
      >
        recent
         to old
      </option>
    </TextField>
  </WithStyles(Grid)>
</WithStyles(Grid)>
`;
